sockeye/decoder.py

```diff
@@ -202,14 +206,13 @@ def init_state_from_encoder(self,
             states = [step, encoder_outputs, source_mask]
 
         batch_size = encoder_outputs.shape[0]
-        # shape: (batch, heads, length, depth_per_head)
-        self_att_key_value_dummies = [mx.nd.zeros((batch_size,
-                                                   self.config.attention_heads,
-                                                   1,
-                                                   self.config.model_size // self.config.attention_heads),
-                                                  ctx=encoder_outputs.context,
-                                                  dtype=encoder_outputs.dtype)] * self.config.num_layers * 2
-        states += self_att_key_value_dummies
+        autoregr_state_shape = self.layers[0].get_states_shape(batch_size)
+        dummy_autoregr_states = [mx.nd.zeros(autoregr_state_shape,
+                                             ctx=encoder_outputs.context,
+                                             dtype=encoder_outputs.dtype)
+                                 for _ in range(self.config.num_layers * self.layers[0].num_state_tensors)]
+
+        states += dummy_autoregr_states
```

Comment:
I think we should either check that `num_state_tensors` is the same across layers or directly use individual layers instead of `layers[0]`:
```suggestion
        dummy_autoregr_states = [mx.nd.zeros(layer.get_states_shape(batch_size),
                                             ctx=encoder_outputs.context,
                                             dtype=encoder_outputs.dtype)
                                 for layer in self.layers for _ in range(0, layer.num_state_tensors)]

        states += dummy_autoregr_states
```

Reply:
Definitely, I missed that.